File: CLOCKSYNC.java
import java.util.Arrays;
import java.util.Scanner;


public class Clocksync {
    static int INF = 99999;
    static int NUM_SWITCHES = 10;
    static int[][] switchInfo = {
            {0, 1, 2},              // SWITCH 0
            {3, 7, 9, 11},          // SWITCH 1
            {4, 10, 14, 15},        // SWITCH 2
            {0, 4, 5, 6, 7},        // SWITCH 3
            {6, 7, 8, 10, 12},      // SWITCH 4
            {0, 2, 14, 15},         // SWITCH 5
            {3, 14, 15},            // SWITCH 6
            {4, 5, 7, 14, 15},      // SWITCH 7
            {1, 2, 3, 4, 5},        // SWITCH 8
            {3, 4, 5, 9, 13}        // SWITCH 9
    };

    static boolean areAligned(int[] clocks) {
        for (int time : clocks) {
            if (time != 12) {
                return false;
            }
        }
        return true;
    }

    static void pushSwitch(int[] clocks, int switchIdx) {
        for (int clockIdx : switchInfo[switchIdx]) {
            clocks[clockIdx] += 3;
            if (clocks[clockIdx] == 15) {
                clocks[clockIdx] = 3;
            }
        }
    }

    static int solve(int[] clocks, int switchIdx) {
        if (switchIdx == NUM_SWITCHES) {
            return areAligned(clocks) ? 0 : INF;
        }
        int ret = INF;
        for (int count = 0; count < 4; count++) {
            ret = Math.min(ret, count + solve(clocks, switchIdx + 1));
            pushSwitch(clocks, switchIdx);
        }
        return ret;
    }


    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int num_prob = Integer.parseInt(scanner.nextLine());
        while (num_prob-- > 0) {
            String[] line = scanner.nextLine().split(" ");
            int[] clocks = Arrays.stream(line).mapToInt(Integer::parseInt).toArray();
            int ret = solve(clocks, 0);
            System.out.println(ret < INF ? ret : -1);
        }
        scanner.close();
    }
}

File: CLOCKSYNC.py
switch_info = '''
0   0, 1, 2
1   3, 7, 9, 11
2   4, 10, 14, 15
3   0, 4, 5, 6, 7
4   6, 7, 8, 10, 12
5   0, 2, 14, 15
6   3, 14, 15
7   4, 5, 7, 14, 15
8   1, 2, 3, 4, 5
9   3, 4, 5, 9, 13
'''
switch_list = [None] * 10
for item in switch_info.split('\n')[1:-1]:
    switch_idx = int(item[0])
    switch_target = [int(i) for i in item[1:].split(',')]
    switch_list[switch_idx] = switch_target


INF = 99999
SWITCHES = 10


def are_aligned(clocks):
    if all(clock % 4 == 0 for clock in clocks):
        return True
    return False


def push_switch(clocks, switch_idx):
    for clock_idx in switch_list[switch_idx]:
        clocks[clock_idx] += 1


def solve(clocks, switch_idx=0):
    if switch_idx == SWITCHES:
        return 0 if are_aligned(clocks) else INF

    ret = INF
    for count in range(4):
        ret = min(ret, count + solve(clocks, switch_idx + 1))
        push_switch(clocks, switch_idx)
    return ret


num_prob = int(input())
for prob in range(num_prob):
    clocks = [int(i) / 3 for i in input().split()]
    result = solve(clocks)
    print(result if result < INF else -1)
